### ISSUE TITLE
compile_threads leaking memory

### ISSUE THREAD

**Issue by angusl-gr** [pytorch/pytorch#105832]
### 🐛 Describe the bug

Hi,

I'm seeing a big constant increase in memory usage for the lifetime of training jobs when I perform model compilation (around 7GB to 16GB in this example, much more for larger jobs). From running some profiling tools it appears that 8 `compile_threads` (actually forked processes?) are created and stick around for the lifetime of the job. My suspicion is that these processes are somehow holding onto references to the data that exists in memory in the main process, perhaps through some forking copy-on-write weirdness. I've managed to work around the issue by explicitly performing
```
import torch._inductor.config
torch._inductor.config.compile_threads = 1
```
which seems to force the compilation to happens synchronously and cleans up the process once it's complete, but it feels hacky and presumably slows down the compilation. I've seen the issue on both V100s and A100s.

Any thoughts on why this is happening and how it could be fixed would be appreciated, many thanks!

Here is a reproducer:
```
import numpy as np
import torch
from torch.utils.data import DataLoader

# uncomment to fix memory usage during compilation
# import torch._inductor.config
# torch._inductor.config.compile_threads = 1


class TestModel(torch.nn.Module):
    def __init__(self, d_model: int):
        super().__init__()
        self.model = torch.nn.Sequential(
            torch.nn.Linear(d_model, d_model),
            torch.nn.ReLU(),
            torch.nn.Linear(d_model, d_model),
            torch.nn.ReLU(),
            torch.nn.Linear(d_model, d_model),
        )

    def forward(self, x):
        return self.model(x)


def run() -> None:
    model = TestModel(d_model=512).cuda()
    model = torch.compile(model)

    for epoch in range(10):
        data = np.random.rand(2**14, 10, 512).astype(np.float32)
        dl = DataLoader(data, batch_size=256, num_workers=4)
        for batch in dl:
            batch = batch.cuda()
            out = model(batch)


if __name__ == "__main__":
    run()
```

### Versions

```
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-121-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla V100-SXM2-32GB
Nvidia driver version: 515.48.07
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] flake8==5.0.4
[pip3] mypy==1.3.0
[pip3] mypy-extensions==1.0.0
[pip3] mypy-protobuf==3.4.0
[pip3] numpy==1.21.4
[pip3] pytorch-lightning==2.0.4
[pip3] pytorch-triton==2.0.0+0d7e753227
[pip3] pytorch-triton-rocm==2.0.1
[pip3] torch==2.0.1
[pip3] torcharrow==0.1.0
[pip3] torchdata==0.6.1
[pip3] torchmetrics==0.10.2
[pip3] triton==2.0.0
[conda] Could not collect
```

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

**Comment by shunting314** [comment]
> I'm seeing a big constant increase in memory usage for the lifetime of training jobs when I perform model compilation (around 7GB to 16GB in this example, much more for larger jobs).

Is the usage increase for CPU memory or GPU memory?

**Comment by angusl-gr** [comment]
CPU memory

**Comment by shunting314** [comment]
ok. Can you also share how you measure the increased CPU memory usage? This can help people repro the problem and confirm a fix. I assume trivially measure the memory usage of the parent process does not work well since we need count the memory usage of forked child processes.

**Comment by angusl-gr** [comment]
I used the [memory-profiler](https://pypi.org/project/memory-profiler/) Python library, specifically with the `--include-children` and `--multiprocess` flags, to measure the memory consumption of subprocesses and to also report the total memory usage. 

So something like:
`mprof run train.py --include-children --multiprocess`
followed by
`mprof plot`.
to see the results.

I have examples of the resulting plots I would like to share but it is difficult to move them outside of my employer's intranet, I will see what I can do.

**Comment by shunting314** [comment]
This one line patch seems to work:

```
diff --git a/torch/_inductor/codecache.py b/torch/_inductor/codecache.py
index 1395bbab671..2c948e1ee76 100644
--- a/torch/_inductor/codecache.py
+++ b/torch/_inductor/codecache.py
@@ -1300,6 +1300,7 @@ class AsyncCompile:
                     scope[key] = result.result()
                     pbar.update(1)

+        self.pool().shutdown()
         _compile_end()
```

If you use a pytorch built from source, can you try this out?

**Comment by shunting314** [comment]
> (I had to use self.process_pool() but I assume this is what you meant)

Ah, yea, thanks for pointing out.

> However I do still see a big memory spike (up to around 15GB) at the start of the job while all the compilation processes are alive.

Hm, I think it's expected since different child processes compile different triton kernels in parallel. Each child process should consume some dedicated resource/memory even if it's forked from the parent in a Copy-On-Write way. You only see increased memory usage at the beginning of the job since the parallel compiling happens at the early stage (when the compiled wrapper is being loaded.)

Right now we decide the  process pool size for async compiling by number of CPUs (capped by 32). The thing you already tried `torch._inductor.config.compile_threads = 1` disabled the parallel compiling. But if you want to reduce the memory usage for async compiling while still keep doing compiling in parallel, you can set the config to something larger than 1 but smaller than the default. There is also an environment variable 'TORCHINDUCTOR_COMPILE_THREADS' to control this.

**Comment by shunting314** [comment]
@angusl-gr BTW, it's not safe to shutdown the process pool inside inductor (`class AsyncCompile`). After compiling one graph we may need compile another graph in at least the following cases
- compiling backward graph for training
- compiling other graphs due to graph break
- compiling more graphs due to dynamic shape
- user call torch.compile to compile another model
- etc.

We can not shutdown the process group before we are guaranteed that there will not be more graphs to compile.

Inductor does not have this knowledge. The decision should be made by the application. If the application writer are sure no more compilation will happen, doing the following thing will avoid the host side memory leak.
- do a warm up run to trigger the real compilation
- shutdown the compilation process pool.

In you specific case, add the following code in your script:
```
    model(torch.rand(256, 10, 512).cuda())
    from torch._inductor.codecache import AsyncCompile
    AsyncCompile.process_pool().shutdown()
```

**Comment by shunting314** [comment]
I'll close the issue now. Free free to re-open it if you have suggestions to let inductor do better to handle such memory leak.

**Comment by pratogab** [comment]
I get a CPU memory leak with the following code (torch v2.0.1):
```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM


def main():
    model = AutoModelForCausalLM.from_pretrained('facebook/opt-125m').to("cuda")
    tokenizer = AutoTokenizer.from_pretrained('facebook/opt-125m')

    model = torch.compile(model, backend="inductor")

    while True:
        x = tokenizer(['I like turtles.'], return_tensors='pt')
        x = x.to("cuda")
        output = model(**x)

if __name__ == '__main__':
    main()
```
If I am not mistaken, this seems to be the same problem. @shunting314 you wrote:
> You only see increased memory usage at the beginning of the job since the parallel compiling happens at the early stage (when the compiled wrapper is being loaded.)

I see a memory leak throughout the entire life of the process, until I run out-of-memory. Could you detail the solution that is proposed?

Also, isn't torch 2's promise that users only need to add one line: `model = torch.compile(model)`?

**Comment by shunting314** [comment]
@pratogab  how long do you see OOM after running the script and how much CPU memory your machine has (I assume you run out of CPU memory rather than GPU memory)? From a first look, it's a different issue. The issue we discusses here (actually it's by design) is we have created a few threads to do parellel compiling. Those threads consume memory and without killing those threads, those memory are not released. However we should not see OOM due to this since the memory usage should be fixed and not that large to causing OOM. 

Any yes, by default you should not need extra code change and applying `torch.compile` is enough. What I mentioned above is, if you really want to reclaim some CPU memory, you can by adding some extra work. This is usually not necessary though.

(BTW, reply late since I was on leave and just saw your message)

**Comment by shunting314** [comment]
@pratogab  I ran you script for 17K iterations so far and the memory usage is quite stable. It can be some issue we have already fixed. You may try pytorch nightly.

Also the memory usage discussed in this issue can be avoided by disabling parallel compiling (which results in longer compilation time). Here is the envvar to disable parallel compiling: `TORCHINDUCTOR_COMPILE_THREADS=1 python ...` . You mean try this to rule out the extra memory usage from parallel compiling

**Comment by pratogab** [comment]
The issue indeed seems to have been fixed in torch 2.1. I was using torch 2.0.1 beforehand. Thank you very much!
